File: Olympiad/COCI/official/2008/contest4_solutions/veci.cpp
#include <cstdio>

int freq[10];

int same_digits( int x, int y ) {
   for( int i = 0; i < 10; ++i ) freq[i] = 0;

   while( x > 0 ) { ++freq[x%10]; x /= 10; }
   while( y > 0 ) { --freq[y%10]; y /= 10; }

   for( int i = 0; i < 10; ++i )
      if( freq[i] != 0 ) return 0;

   return 1;
}

int main( void ) {
   int x, y;
   scanf( "%d", &x );

   for( y = x+1; y-x < 1000000 && !same_digits( x, y ); ++y );

   if( y-x >= 1000000 ) printf( "0\n" );
   else printf( "%d\n", y );

   return 0;
}

File: Olympiad/COCI/official/2008/contest4_solutions/veci.pas
program veci;

var
   x, y : longint;
   freq : array[0..9] of longint;

function sameDigits( x, y : longint ) : boolean;
var
   i : longint;
begin
   for i := 0 to 9 do freq[i] := 0;

   while x > 0 do begin inc(freq[x mod 10]); x := x div 10; end;
   while y > 0 do begin dec(freq[y mod 10]); y := y div 10; end;

   sameDigits := true;
   for i := 0 to 9 do
      if freq[i] <> 0 then
         sameDigits := false;
end;

begin
   readln( x );

   y := x+1;
   while (y-x < 1000000) and not sameDigits( x, y ) do inc( y );

   if y-x >= 1000000 then writeln( '0' ) else writeln( y );
end.
